src/oneD/Flow1D.cpp

```diff
@@ -732,21 +732,25 @@ void Flow1D::evalSpecies(double* x, double* rsd, int* diag,
         for (size_t k = 0; k < m_nsp; k++) {
             sum += Y(x,k,jmin);
             rsd[index(c_offset_Y+k, jmin)] = -(m_flux(k, jmin) +
-                                                rho_u(x, jmin) * Y(x, k, jmin));
+                                               rho_u(x, jmin) * Y(x, k, jmin));
+            // diag[index(c_offset_Y+k, jmin)] are zero and do not need to be reset
         }
+        // Residuals may be modified by left boundary (for k != leftExcessSpecies);
+        // excess species residual is handled as algebraic constraint.
         rsd[index(c_offset_Y + leftExcessSpecies(), jmin)] = 1.0 - sum;
-        diag[index(c_offset_Y + leftExcessSpecies(), jmin)] = 0;
     }
 
     if (jmax == m_points - 1) { // right boundary
         double sum = 0.0;
         for (size_t k = 0; k < m_nsp; k++) {
             sum += Y(x,k,jmax);
-            rsd[index(k+c_offset_Y, jmax)] = m_flux(k, jmax-1) +
+            rsd[index(c_offset_Y+k, jmax)] = m_flux(k, jmax-1) +
                                              rho_u(x, jmax)*Y(x, k, jmax);
+            // diag[index(c_offset_Y+k, jmin)] are zero and do not need to be reset
         }
+        // Residuals may be modified by right boundary (for k != rightExcessSpecies);
+        // excess species residual is handled as algebraic constraint.
```

Comment:
Thanks for the updates here @ischoegl! I think it helps a bit to have a note about what is happening here, but the "what" is also clear from the code itself IMO. I think it would even better to have a note about _why_ we can treat the excess species with an algebraic constraint (which in other situations may result in overdetermined systems, see e.g., the species production equations where we allow small negative values rather than enforcing 1-sum(k-1)). I think a link to documentation about the why would suffice, if this is already explained some where. Recognizing as well that we don't need this to become a textbook on numerical methods... Thanks!

Reply:
Thanks, @bryanwweber. I guess my point was mostly to re-introduce some older comments from `StFlow.cpp` (which were not preserved in the refactoring that took place recently):
```
            // The default boundary condition for species is zero flux. However,
            // the boundary object may modify this.
```
It is true that part of this comment was moved to the docstring of the method, but I don't think is sufficiently clear when reading this implementation code that residuals are later modified. The diag setters were only partially implemented but also superfluous (so reminding users that there is no time stepping involved is imho a better option). The algebraic constraint for mass fraction is imho self-evident, but I guess mileage varies? Overall, I tried to improve code that is not straight-forward, and if you have a better suggestion, I'd be happy to adopt.

Reply:
The only suggestion I have is to explain (either here or on the website) why we can use an algebraic constraint for the residual here, specifically because I feel fairly confident that in other cases we can't. I don't know the answer, hence I assume others won't either. My comment was that the code already is clear that this is an algebraic constraint, so adding a comment which states that fact explicitly is of limited utility IMO. The why is far more important IMO.

Reply:
Fair point, but since this is de facto legacy code, it's hard to guess. My intuition tells me that it may have something to do with the fact that these are boundary conditions and thus have to be treated differently than interior points, where we don't have that imposed constraint. Will remove the second commit as apparently this isn't an improvement over the status quo and I don't have time to dig into the theory.